book-content/chapters/15-designing-your-types.md

```diff
@@ -0,0 +1,1444 @@
+
+### Multiple Type Parameters
+
+Generic types can accept multiple type parameters, allowing for even more flexibility.
+
+We could expand the `ResourceStatus` type to include a second type parameter that represents metadata that accompanies the resource:
+
+```tsx
+type ResourceStatus<TContent, TMetadata> =
+  | {
+      status: "available";
+      content: TContent;
+      metadata: TMetadata;
+    }
+  | {
+      status: "unavailable";
+      reason: string;
+    };
+```
+
+Now we can define the `StreamingPlaylist` and `StreamingAlbum` types, we can include metadata specific to each resource:
+
+```tsx
+type StreamingPlaylist = ResourceStatus<
+  {
+    id: number;
+    name: string;
+    tracks: string[];
+  },
+  {
+    creator: string;
+    artwork: string;
+    dateUpdated: Date;
+  }
+>;
+
+type StreamingAlbum = ResourceStatus<
+  {
+    id: number;
+    title: string;
+    artist: string;
+    tracks: string[];
+  },
+  {
+    recordLabel: string;
+    upc: string;
+    yearOfRelease: number;
+  }
+>;
+```
+
+Like before, each type maintains the same structure defined in `ResourceStatus`, but with its own content and metadata.
+
+You can use as many type parameters as you need in a generic type. But just like with functions, the more parameters you have, the more complex your types can become.
+
+### All Type Arguments Must Be Provided
+
+What happens if we don't pass a type argument to a generic type? Let's try it with the `ResourceStatus` type:
+
+```ts twoslash
+// @errors: 2314
+type ResourceStatus<TContent, TMetadata> =
+  | {
+      status: "available";
+      content: TContent;
+      metadata: TMetadata;
+    }
+  | {
+      status: "unavailable";
+      reason: string;
+    };
+// ---cut---
+type Example = ResourceStatus;
+```
+
+TypeScript shows an error that tells us that `ResourceStatus` requires two type arguments. This is because by default, all generic types _require_ their type arguments to be passed in, just like runtime functions.
+
+### Default Type Parameters
+
+In some cases, you may want to provide default types for generic type parameters. Like with functions, you can use the `=` to assign a default value.
+
+By setting `TMetadata`'s default value to an empty, we can essentially make `TMetadata` optional:
+
+```tsx
+type ResourceStatus<TContent, TMetadata = {}> =
+  | {
+      status: "available";
+      content: TContent;
+      metadata: TMetadata;
+    }
+  | {
+      status: "unavailable";
+      reason: string;
+    };
+```
+
+Now, we can create a `StreamingPlaylist` type without providing a `TMetadata` type argument:
+
+```tsx
+type StreamingPlaylist = ResourceStatus<{
+  id: number;
+  name: string;
+  tracks: string[];
+}>;
+```
+
+If we hover over it, we'll see that it's typed as expected, with `metadata` being an empty object:
```

Comment:
given that you call out in earlier content that `{}` doesn't mean an empty object you probably shouldn't call it an empty object here :p but yes, I don't know how to name it better